### ISSUE TITLE
A strange phenomenon when gp-lstm is applied to a simulated dataset

### ISSUE THREAD

**Issue by WilesDai** [alshedivat/keras-gp#27]
Many thanks for your patience to read my question!

I have constructed  a x sequence and y sequence:
X = np.linspace(0, 1, 2000)
y[i] = 0.5 * X[i] + 2 + np.random.normal(0, X[i])

However, when I applied the gp-lstm model to my simulated data, the prediction of mean sequence and volatility sequence are both constant sequence.

predicted mean  sequence
[[[4.63309287]
  [4.63309287]
  [4.63309287]
  [4.63309287]
  [4.63309287]
  [4.63309287]
.....................
  [4.63309287]
  [4.63309287]
  [4.63309287]
  [4.63309287]
  [4.63309287]
  [4.63309287]
  [4.63309287]
  [4.63309287]
  [4.63309287]]]

predicted volatility  sequence
[[[0.01834713]
  [0.01834713]
  [0.01834713]
  [0.01834713]
  [0.01834713]
  [0.01834713]
  [0.01834713]
 ....................
  [0.01834713]
  [0.01834713]
  [0.01834713]
  [0.01834713]
  [0.01834713]
  [0.01834713]]]

**Comment by WilesDai** [comment]
Actually, I have tried several time series datasets,  no matter the data is normalized or not，the prediction of mean  sequence and volatility sequence  all keep constant over time. That's very strange! I don't know if I have ignored any step when I utilize the gp-lstm model and model these datasets.  I would appreciate it very much if you could give me some advice!

**Comment by WilesDai** [comment]
The following is the code.

```
import numpy as np
np.random.seed(42)
from keras.callbacks import EarlyStopping
from keras.layers import Input, SimpleRNN, LSTM
from keras.optimizers import Adam

from kgp.layers import GP
from kgp.models import Model
from kgp.utils.experiment import train
from kgp.losses import gen_gp_loss


def gp_lstm(timesteps, batch_size, data):
    input_shape = (timesteps, len(data["train"][0][0, 0, :]))  # time steps, 2 dimensions
    print("input_shape", input_shape)
    nb_train_samples = len(data["train"][0])
    gp_hypers = {'lik': -2.0, 'cov': [[-0.7], [0.0]]}
    inputs = Input(shape=input_shape)

    rnn = LSTM(10)(inputs)
    gp = GP(gp_hypers,
            batch_size=batch_size,
            nb_train_samples=nb_train_samples
            )
    outputs = [gp(rnn)]
    model = Model(inputs=inputs, outputs=outputs)

    # Compile the model
    loss = [gen_gp_loss(gp) for gp in model.output_layers]
    model.compile(optimizer=Adam(1e-1), loss=loss)
    model.summary()
    callbacks = [EarlyStopping(monitor='val_mse', patience=20)]
    epochs = 1000
    history = train(model, data, callbacks=callbacks, epochs=epochs, batch_size=batch_size, verbose=2)
    X_test, y_test = data['test']
    preds = model.predict(X_test, return_var=True)
    y_preds_array = np.array(list(preds))
    y_preds_mean = y_preds_array[0, :, :, :]
    y_preds_volatility = y_preds_array[1, :, :, :]
    return y_preds_mean, y_preds_volatility


# input and output data construction
def data_process(feature, target, timesteps):
    feature_seq = []
    target_seq = []

    for j in range(timesteps, len(feature)):
        feature_seq.append(feature[(j-timesteps):j, :])
        target_seq.append(target[j, 0])

    feature_seq, target_seq = np.array(feature_seq), np.array(target_seq)
    feature_seq = feature_seq.reshape(-1, timesteps, 1)
    target_seq = target_seq.reshape(-1, 1, 1)

    return feature_seq, target_seq


# instance 1
X = np.linspace(0, 10, 2000)
y_origin = np.zeros(X.shape)
epsilon = np.zeros(X.shape)
y = np.zeros(X.shape)
for i in range(len(y)):
    y_origin[i] = 0.5 * X[i] + 2
    epsilon[i] = np.random.normal(0, X[i])
    y[i] = 0.5 * X[i] + 2 + np.random.normal(0, X[i])


X = X.reshape(-1, 1)
y_origin = y_origin.reshape(-1, 1)
epsilon = epsilon.reshape(-1, 1)
y = y.reshape(-1, 1)


time_steps = 2
X_seq, y_seq = data_process(X, y, time_steps)
test_start = len(X_seq)-300
train_end = int(0.7*(len(X_seq)-300))

X_train, y_train = X_seq[:train_end], y_seq[:train_end]
X_valid, y_valid = X_seq[train_end:test_start], y_seq[train_end:test_start]
X_test, y_test = X_seq[test_start:], y_seq[test_start:]


dataset_constructed = {
    'train': [X_train, y_train],
    'valid': [X_valid, y_valid],
    'test': [X_test, y_test],
}
for set_name in dataset_constructed:
    y = dataset_constructed[set_name][1]
    y = y.reshape((-1, 1, np.prod(y.shape[1:])))
    dataset_constructed[set_name][1] = [y[:, :, i] for i in range(y.shape[2])]


mean, volatility = gp_lstm(time_steps, 128, dataset_constructed)
```
